criu/cr-restore.c

```diff
@@ -2448,6 +2448,21 @@ static int restore_root_task(struct pstree_item *init)
 		pr_err("Unable to flush breakpoints\n");
 
 	finalize_restore();
+	/*
+	 * Some external devices such as GPUs might need a very late
+	 * trigger to kick-off some events, memory notifiers and for
+	 * restarting the previously restored queues during criu restore
+	 * stage. This is needed since criu pie code may shuffle VMAs
+	 * around so things such as registering MMU notifiers (for GPU
+	 * mapped memory) could be done sanely once the pie code hands
+	 * over the control to master process.
+	 */
+	for_each_pstree_item(item) {
+		pr_info("Run late stage hook from criu master for external devices\n");
+		ret = run_plugins(RESUME_DEVICES_LATE, item->pid->real);
```

Comment:
Not familiar enough with plugins myself, but is that correct to try to run the plugin always. More a general question if there isn't a flag which tells CRIU that a plugin exists. Or does `run_plugins()` handle it.

Reply:
I think run_plugin can handle if there is no plugin. This is a unique scenario when we want to give a final chance to all resuming tasks to do some special handling like restarting MMU notifications, events etc because *now* all the children of criu master restorer have their VMAs finalized and tasks are ready to run. In our case, the kernel mode driver makes sure if the pid indeed belongs to kfd and then only starts the previously setup (during criu restore stage) MMU notifies etc.  https://github.com/RadeonOpenCompute/ROCK-Kernel-Driver/commit/7ec3510634fdcdbca0c1f1673f31a4915d748ee2 and https://github.com/RadeonOpenCompute/criu/commit/a9979fabfd023a0fa2e993c350c2f0d451e330cd